recipes/libcublas/meta.yaml

```
@@ -0,0 +1,108 @@
+{% set name = "libcublas" %}
+{% set version = "12.0.1.189" %}
+{% set cuda_version = "12.0" %}
+{% set platform = "linux-x86_64" %}  # [linux64]
+{% set platform = "linux-ppc64le" %}  # [ppc64le]
+{% set platform = "linux-sbsa" %}  # [aarch64]
+{% set platform = "windows-x86_64" %}  # [win]
+{% set extension = "tar.xz" %}  # [not win]
+{% set extension = "zip" %}  # [win]
+
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  url: https://developer.download.nvidia.com/compute/cuda/redist/{{ name }}/{{ platform }}/{{ name }}-{{ platform }}-{{ version }}-archive.{{ extension }}
+  sha256: 38272cc9184b4a57a6250fc2cd914ffc827f9f8cd5a3e1356bc3c31f843caf8f  # [linux64]
+  sha256: 2d46e735f8e76e2eb7df1b4f41d56aa8d5d2683f99988bac82d6ae33fb6d344b  # [ppc64le]
+  sha256: 96b68f6821684f39f338862639ce74a74c75830bf44dcd17ef372a1cb8c2304d  # [aarch64]
+  sha256: 88f29bec81880844da2eee13f1cdc51e1ccd8b4c60af6df633577810b30f80d0  # [win]
+
+build:
+  number: 0
+  skip: true  # [osx]
+
+requirements:
+  build:
+    - {{ compiler("c") }}
+    - {{ compiler("cxx") }}
+    - sysroot_{{ target_platform }} 2.17   # [linux]
+
+outputs:
+  - name: libcublas
+    requirements:
+      host:
+         - cuda-version {{ cuda_version }}
+      run:
+         - {{ pin_compatible("cuda-version", max_pin="x.x") }}
+    files:
+      - lib/lib*blas*.so.*            # [linux]
+      - targets/*/lib/lib*blas*.so.*  # [linux]
+      - Library\bin\cublas*.dll                   # [win]
+      - Library\bin\nvblas*.dll                   # [win]
+    test:
+      commands:
+        - test -f $PREFIX/lib/libcublas.so.*  # [unix]
+        - if not exist %LIBRARY_LIB%\\cublas.lib exit 1  # [win]
+
+  - name: libcublas-dev
+    build:
+      run_exports:
+        - {{ pin_subpackage("libcublas", max_pin="x") }}
+    requirements:
+      host:
+        - cuda-version {{ cuda_version }}
+      run:
+        - {{ pin_compatible("cuda-version", max_pin="x.x") }}
+        - libcublas >={{ version }}
+      run_constrained:
+        - libcublas-static >={{ version }}
```

@jakirkham @leofang I don't know how to constrain `libcublas-static` so that it only matches the current major version, but I feel like that might be desired? `libcublas` has a `pin_subpackage("libcublas", max_pin="x")`. Does this need a `pin_compatible`, or is that possible with `run_constrained`?

---

Or maybe the `run_exports: - {{ pin_subpackage("libcublas", max_pin="x") }}` above isn't needed at all, due to constraints on `cuda-version`? I feel like there's asymmetry here that we don't want, but I'm unsure of how to improve it.

---

For comparison, https://github.com/conda-forge/staged-recipes/pull/21916/ is not defining a `run_exports` on `libcurand-dev`.

---

Right, I thought we could do this?
```yaml
run_constrained:
  - {{ pin_subpackage("libcublas", max_pin="x") }}
```
My understanding is that `pin_compatible` is only used if it's a `build` requirement, which `-static` isn't.

---

We could trying using [`{{ pin_subpackage('libcublas-static', exact=True) }}`]( https://docs.conda.io/projects/conda-build/en/latest/resources/variants.html?highlight=pin_subpackage%20exact#extra-jinja2-functions )

We should have `run_exports` since using dynamic libraries is the common case

Users of static libraries can add [`ignore_run_exports_from`]( https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html?highlight=ignore_run_exports_from#export-runtime-requirements ) to drop the runtime dependency

---

@jakirkham Can you make suggestions for what you're proposing? Just want to make it concrete so I know I'm understanding correctly.

---

Since the `-dev` packages are only used in `host` while building packages, am guessing we want exact alignment between headers, libraries, and other build config files (pkg-config, CMake, etc.). Does that seem reasonable or am I missing something?

---

OK, so like this?
```yaml
run_constrained:
  - {{ pin_subpackage("libcublas-static", exact=True) }}
```

---

Do we want to align `libcublas` too?

```yaml
      run:
        ...
        - {{ pin_subpackage("libcublas", exact=True) }}
      run_constrained:
        - {{ pin_subpackage("libcublas-static", exact=True) }}
```

---

> Since the -dev packages are only used in host while building other packages

I’m not sure if that’s true. For example, libraft-headers will have runtime dependencies on -dev packages, because it needs to install CUDA headers for any downstream libraries using libraft-headers. But I suppose that transitively, libraft-headers is only used in `host`?

Also I guess if you’re making a developer environment that can build CUDA packages from source, you’ll need to install the dev packages, but forcing the same runtime libraries is okay as long as that run requirement is not exported to recipes using the -dev packages (those should export major-compatible pinnings).

---

Sorry, let me take one step back. I don't follow either of you anymore... 😅

@jakirkham, why do we have to pin exactly, instead of just letting `cuda-version` to kick in? I think this is Bradley's original question 
```suggestion
        - libcublas-static
```
and based on the discussion https://github.com/conda-forge/cuda-version-feedstock/issues/1#issuecomment-1462998398, it seems to be what we had agreed upon?

@bdice, I don't understand how is `-dev` packages being used at runtime related to `-static` packages here? This is `run_constrained`, not `run`, so it's OK that the `-dev` package is installed to offer the headers at runtime. It won't bring in the `-static` package unless the user explicitly asks for it. Am I missing something?